devito/types/object.py

```diff
@@ -84,6 +85,177 @@ def function(self):
     __reduce_ex__ = Pickable.__reduce_ex__
 
 
+class AbstractObjectWithShape(Basic, sympy.Basic, Pickable):
+
+    # need to check this?
+    # __rkwargs__ = ('name', 'dtype', 'is_const', 'grid', 'shape', 'dimensions')
+
+    def __new__(cls, *args, **kwargs):
```

Comment:
This still looks like there is a quite large amount of overlap with `AbstractFunction` so I'm not sure why this is needed.

Reply:
@mloubout AbstractObject doesn't inherit from AbstractFunction, so unless we lift some of the construction logic to the closest common ancestor (as in, `Basic`), this needs to stay here, I think

Reply:
But this doesn't need to inheriting from `AbstractObject` if there is already an existing type that does all of it. This just redefines `AbstractFunction` with a new calss name and `_const` it should be a 2 liners inheriting from `AbstractFunction`